openassessment/fileupload/backends/base.py

```diff
@@ -0,0 +1,119 @@
+from django.conf import settings
+
+from ..exceptions import FileUploadInternalError
+from ..exceptions import FileUploadRequestError
+
+
+class Settings(object):
+    """Store settings related to file upload
+
+    The following settings are used:
+
+        FILE_UPLOAD_STORAGE_BUCKET_NAME (str, required): name of the bucket
+        (AWS or local file directory) to which content will be uploaded.
+
+        FILE_STORAGE_SETTING (str, defaults to "submissions_attachment"): this
+        will be used to prefix all stored file names. The specified file
+        prefix for the storage must be publicly viewable or all uploaded images
+        will not be seen.
+    """
+
+    @classmethod
+    def get_bucket_name(cls):
+        bucket_name = getattr(settings, "FILE_UPLOAD_STORAGE_BUCKET_NAME", None)
+        if not bucket_name:
+            raise FileUploadInternalError("No bucket name configured for FileUpload Service.")
+        return bucket_name
+
+    @classmethod
+    def get_prefix(cls):
+        return getattr(settings, "FILE_UPLOAD_STORAGE_PREFIX", "submissions_attachments")
+
+
+class BaseBackend(object):
+
+    # Time (in seconds) before an upload url expires
+    UPLOAD_URL_TIMEOUT = 3600
+
+    # Time (in seconds) before a download url expires
+    DOWNLOAD_URL_TIMEOUT = 1000
+
+    def get_upload_url(self, key, content_type):
+        """Request a one-time upload URL to upload files.
+
+        Requests a URL for a one-time file upload.
+
+        Args:
+            key (str): A unique identifier used to construct the upload location and
+                later, can be used to retrieve the same information. This service
+                must be able to identify data for both upload and download using
+                this key.
+            content_type (str): The content type for the file.
+
+        Returns:
+            A URL (str) to use for a one-time upload.
+
+        Raises:
+            FileUploadInternalError: Raised when an internal error occurs while
+                retrieving a one-time URL.
+            FileUploadRequestError: Raised when the request failed due to
+                request restrictions
+
+        """
+        raise NotImplementedError
```

Comment:
Nit: Consider using the Python `abc` module to make these abstract methods, since every resulting Backend deriving from this class needs to implement them. It'd make the error happen upon instantiation instead of on calling the method.